fhir-persistence/src/main/java/com/ibm/fhir/persistence/ResourceResult.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * (C) Copyright IBM Corp. 2021, 2022
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.persistence;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.ibm.fhir.model.annotation.Required;
+import com.ibm.fhir.model.resource.Resource;
+
+/**
+ * The base wrapper result wrapper used to represent a resource being returned from a
+ * persistence interaction.
+ * Instances are immutable and can be constructed via {@code new ResourceResult.Builder<T>()}.
+ */
+public class ResourceResult<T extends Resource> {
```

Comment:
random thought: I wonder if this class could help us implement #2432 in the future.
is there anything we'd design differently?

Reply:
Interesting, yes I think so. I am also hoping to use it to improve offload performance for reads which could be asynchronous. It also will allow us to speed up ingestion where we do a read but don't actually need the resource itself, just the metadata.